Setup instructions revisions (#14)
OpenLightingProject/website: download_and_install/compiling_from_source.md

Compiling From Source
=====================

> OLA source code lives on [GitHub](https://github.com/OpenLightingProject/ola). Either download a source tarball or

Reviewer: ... download a source tarball from the [OLA Releases](https://github.com/OpenLightingProject/ola/releases) page or ...

Author: Rephrased slightly to separate stable release downloads from bleeding-edge.

run `git clone https://github.com/OpenLightingProject/ola.git`.

Dependencies
============

You need a couple of libraries installed for everything to work correctly.

First you’ll need at least the following:

> - [avahi](https://www.avahi.org) if you want discovery enabled.

Reviewer: `discovery` -> `network discovery` ?

Author: Yep.  Fixed the bad link to protobuf as well.  Most of this page came directly from the existing version, with only the assorted installation commands updated.  That'll teach me!

- [avahi](https://www.avahi.org) if you want to use network discovery.
- [microhttpd](https://www.gnu.org/software/libmicrohttpd/) (if you want the web UI).
- autoconf
- automake
- cppunit
- ncurses
- lex (or flex)
- libftdi
- liblo
- libtool
- libusb
- pkg-config
- [Protobuf](https://developers.google.com/protocol-buffers) (version 3.6 or later)
- uuid or ossp uuid
- yacc (or bison)
- zlib

The Python library also requires

- Numpy
- Protobuf for Python

Debian/Ubuntu
-------------

Debian/Ubuntu users can install all dependencies with:

    sudo apt-get install \
        autoconf \
        automake \
        bison \
        flex \
        g++ \
        libavahi-client-dev \
        libcppunit-dev \
        libftdi-dev \
        liblo-dev \
        libmicrohttpd-dev \
        libncurses-dev \
        libprotobuf-dev \
        libprotoc-dev \
        libtool \
        libusb-1.0.0-dev \
        make \

>         pkg-config \

Reviewer 2: Add protobuf-compiler for protoc (on Ubuntu at least)
```suggestion
        protobuf-compiler \
        pkg-config \
```

        autoconf \
        automake \
        avahi-devel \
        bison \
        cppunit-devel \
        flex \
        gcc-c++ \
        libftdi-devel \
        liblo-devel \
        libmicrohttpd-devel \
        libtool \
        libusbx-devel \
        make \
        ncurses-devel \
        protobuf-lite-devel \
        uuid-devel \
        zlib-devel

macOS
-----

Dependencies are available from [MacPorts](https://www.macports.org/):

    sudo port install \
        autoconf \
        automake \
        bison \
        cppunit \
        flex \
        libftdi1 \
        liblo \
        libmicrohttpd \
        libtool \
        libusb \
        libuuid \
        ncurses \
        pkgconfig \
        protobuf3-cpp \
        zlib

or [Homebrew](https://brew.sh/):

    brew install \
        autoconf \
        automake \
        bison \
        cppunit \
        flex \
        libftdi \
        liblo \
        libmicrohttpd \
        libtool \
        libusb \
        libuuid \
        ncurses \
        pkg-config \
        protobuf \
        zlib

Python dependencies
-------------------

If you wish to build the Python library, additional libraries are required:

>     pip3 install --user \
>         numpy \
>         protobuf

Reviewer 2: This seems to cause issue on some Ubuntu flavours (I think they end up in a different path and not found by configure), instead use:
```suggestion
    sudo apt-get install \
        python-numpy \
        python-protobuf
```

Or the python3 versions of them...

Author: Huh, I guess my local Ubuntu might be too new.  I'll check other distros for this issue - I was trying to keep Python with Python, so to speak, by using pip instead of the system package manager.

Reviewer 2: Possibly, there's also the Python 2/3 thing, you can check some of the Travis runs and config history for what I had to do to make it work. This failure also matched what a user on IRC experienced.

Author: Alright, it looks like the Python ClientWrapperTest is failing on Ubuntu when using pip, but passing when using system packages.  It works either way on Fedora.  I'm baffled.  Anyway, updating the Python lib to note this problem.

Reviewer 2: > Alright, it looks like the Python ClientWrapperTest is failing on Ubuntu when using pip, but passing when using system packages. It works either way on Fedora. I'm baffled. Anyway, updating the Python lib to note this problem.

Can you open an issue for this in OLA @danielskeenan as it would be good to get to the bottom of it. Some OS version and test log failure stuff would be useful too.

Author: Will do next time I flip back to Linux, where all of my test VMs are.

Author: Opened OpenLightingProject/ola#1714.  It turned out to be more interesting than I thought!

If the pip3 command does not exist, install python3-pip from your package manager and try again.

Build
=====

If this is the first time compiling with this download, generate the build system:

    autoreconf -i

OLA uses autotools for building. Run

     ./configure --help

> to see all options. The most popular option is `--enable-python-libs` to build the Python Client Module. If you want to

Reviewer: `client` and `module` with non-capitals?

use the RDM responder tests, add `--enable-rdm-tests`.

Once you’ve decided on the options, it’s time to build OLA.

    ./configure <your options>
    make -j$(nproc)
    make check
    sudo make install

Mac users should run use `make -j$(sysctl -n hw.logicalcpu)` instead of `make -j$(nproc)`.

> Finally, run `sudo ldconfig`to make new libraries available.

Reviewer: space missing before "to"

> Compiling From Source

Reviewer 2: Is it worth a bit of an introduction here pointing people at the package instructions where possible (in case they hit these instructions via search)?

It strikes me they probably only want to compile from source if:
- It's not available on their OS
- They want to fix something/test a fix
- They want to try the bleeding edge.

=====================

OLA source code lives on [GitHub](https://github.com/OpenLightingProject/ola).
Stable releases are available from the [releases page](https://github.com/OpenLightingProject/ola/releases);

> for bleeding-edge development releases, run `git clone https://github.com/OpenLightingProject/ola.git`.

Reviewer 2: We probably want to mention branches here too, depending on what they're doing they may want master or a stable parent branch (e.g. 0.10 currently).

Dependencies
============

> You need a couple of libraries installed for everything to work correctly.

Reviewer 2: Couple is probably optimistic :laughing:

    sudo port install \
        autoconf \
        automake \
        bison \
        cppunit \
        flex \
        libftdi1 \
        liblo \
        libmicrohttpd \
        libtool \
        libusb \
        libuuid \
        ncurses \
        pkgconfig \
        protobuf3-cpp \
        zlib

or [Homebrew](https://brew.sh/):

    brew install \
        autoconf \
        automake \
        bison \
        cppunit \
        flex \
        libftdi \
        liblo \
        libmicrohttpd \
        libtool \
        libusb \
        libuuid \
        ncurses \
        pkg-config \
        protobuf \
        zlib

Python dependencies
-------------------

If you wish to build the Python library, additional libraries are required:

    pip3 install --user \
        numpy \
        protobuf

If the pip3 command does not exist, install python3-pip from your package manager and try again.

Build
=====

If this is the first time compiling with this download, generate the build system:

    autoreconf -i

OLA uses autotools for building. Run

     ./configure --help

to see all options. The most popular option is `--enable-python-libs` to build the Python client module. If you want to

> use the RDM responder tests, add `--enable-rdm-tests`.

Reviewer 2: It might be worth mentioning `--disable-all-plugins --enable-foo` which I recommend if you only need the foo plugin and you're compiling on something slow like an older Pi.

Author: Once I've figured out the Python shenanigans, I might rework this section.  Right now, it's copied more-or-less verbatim from the old text.  I am wary of getting too detailed; don't want to risk letting the docs and code get badly out of sync again.

Reviewer 2: It's unlikely the code will change in this sense, we added the `--disable-all-plugins` to configure to make things like this easier. The `--enable-foo` would be fetched from `./configure --help` depending on which plugin you actually wanted to use.

Backwards compatibility wise, if we added an e.g. --only-foo (which I doubt we would, as realistically you probably want `--only-e131 --only-artnet` which name wise doesn't really make sense, we'd keep the old options for compatibility too).

> Once you’ve decided on the options, it’s time to build OLA.

Reviewer 2: They need to uninstall the packaged OLA too or that causes issues as well.

    ./configure <your options>
    make -j$(nproc)
    make check
    sudo make install

Mac users should run use `make -j$(sysctl -n hw.logicalcpu)` instead of `make -j$(nproc)`.

Finally, run `sudo ldconfig` to make new libraries available.

Device drivers
==============

Note that, for some devices, it is necessary to install drivers for OLA to work with them. For example, the
[Open DMX USB](http://opendmx.net/index.php/Open_DMX_USB "Open DMX USB") device needs an additional kernel module that
could be built using the instructions on
[LLA\_and\_Q\_Light\_Controller\_Ubuntu\_Tutorial](http://opendmx.net/index.php/LLA_and_Q_Light_Controller_Ubuntu_Tutorial)

> . For other devices, refer to the corresponding device page on the wiki.

Reviewer 2: Nowadays using the FTDI plugin is probably easier.

Author: This section is also copied verbatim from the old text.  I've never played with FTDI functionality, so may need to do some research here.  I'm all for removing unnecessary references to the often-outdated wiki.

Reviewer 2: See https://www.openlighting.org/ola/get-help/ola-faq/#What_is_the_difference_between_the_different_USB_plugins

I suspect the kernel plugin would in theory be more efficient, but for most people it's probably not worth the hassle of getting that going to.

    ./libtool: line 464: CDPATH: command not found
    /Users/simonn/lighting/lla/libltdl/libtool: line 464: CDPATH: command not found
    /Users/simonn/lighting/lla/libltdl/libtool: line 1142: func_opt_split: command not found
    libtool: Version mismatch error.  This is libtool 2.2.6, but the
    libtool: definition of this LT_INIT comes from an older release.
    libtool: You should recreate aclocal.m4 with macros from libtool 2.2.6
    libtool: and run autoconf again.

Your system uses a different version of libtool. Run:

     libtoolize --ltdl -c -f

and then start from the autoreconf step again.

If you should get the following error try to fix it with one of
[two available solutions](https://groups.google.com/group/open-lighting/msg/72060f6327d30df6):

    Rpc.pb.cc: In copy constructor 'ola::rpc::RpcMessage::RpcMessage(const ola::rpc::RpcMessage&)': 
    Rpc.pb.cc:143: error: base class 'class google::protobuf::Message' should be explicitly initialized in the copy constructor

You should be able to prevent this
by [editing `./src/Makefile.am`](https://groups.google.com/group/open-lighting/msg/c6d86d03dd74ed5b), removing `-Werror`

> and then start from the autoreconfig step again.

Reviewer 2: I think this can just be adding `--disable-fatal-warnings` to configure args now.

Compiling From Source
=====================

OLA developers recommend using the latest [pre-built packages](download_and_install). Compile from source if a release
is not available for your system, you wish to try bleeding-edge features, or want to contribute to development.

**Before compiling and installing from source, be sure to remove any pre-packaged OLA installations, as they will likely
conflict.**

If you're testing development releases and find an issue, please report it on
the [issue tracker](https://github.com/OpenLightingProject/ola/issues).

OLA source code lives on [GitHub](https://github.com/OpenLightingProject/ola). Stable releases are available from
the [releases page](https://github.com/OpenLightingProject/ola/releases); this is what you want if compiling only
because OLA is not packaged on your system. For the latest code,
run `git clone https://github.com/OpenLightingProject/ola.git`.

OLA uses `MAJOR.MINOR.PATCH` branches. The latest development happens on the `master` branch. Features and fixes under

> testing are in the minor branch (i.e. `0.x`). Release candidates are on the patch branch (i.e. `0.10.x`). Checkout your

Reviewer 2: We don't currently do RCs, I don't think there's really enough changes for it, so we just generate 0.10.8 at the point of release.

preferred level of stability with `git checkout <branch>` or a final release with `git checkout tags/<release>`. If

> you're a developer, always work against the `master` branch. See `README.developer` for more developer advice.

Reviewer 2: ```suggestion
you're a developer, work against the `master` branch for new features or the relevant minor branch for bug fixes. See `README.developer` for more developer advice.
```

    autoreconf -i

OLA uses autotools for building. Run `./configure --help` to see all options.

If building OLA for a specific purpose, or on a low-power system such as a Raspberry Pi, consider
using `--disable-all-plugins` and enabling only the specific plugins you plan on using. For example:

    ./configure --disable-all-plugins --enable-e131 --enable-osc

will build only the E1.31 (sACN) and OSC plugins, saving significant build time.

Other popular options are `--enable-python-libs` for using the Python client library, or `--enable-rdm-tests` for the
RDM responder tests.

Once you’ve decided on the options, it’s time to build OLA.

    ./configure <your options>
    make -j$(nproc)
    make check
    sudo make install

Mac users should use `make -j$(sysctl -n hw.logicalcpu)` instead of `make -j$(nproc)`.

Finally, run `sudo ldconfig` to make new libraries available.

Device configuration
====================

Some devices (particularly those that connect via USB) require specific configuration to function.
See [here](https://www.openlighting.org/ola/getting-started/device-specific-configuration/) for instructions on

> configuration common devices.

Reviewer 2: ```suggestion
configuring common devices.
```

     libtoolize --ltdl -c -f

and then start from the autoreconf step again.

If you should get the following error try to fix it with one of
[two available solutions](https://groups.google.com/group/open-lighting/msg/72060f6327d30df6):

    Rpc.pb.cc: In copy constructor 'ola::rpc::RpcMessage::RpcMessage(const ola::rpc::RpcMessage&)': 
    Rpc.pb.cc:143: error: base class 'class google::protobuf::Message' should be explicitly initialized in the copy constructor

try using the `--disable-fatal-warnings` option.

If you get

    In file included from ./common/rdm/PidStoreLoader.h:32:0,
                     from common/rdm/PidStore.cpp:24:
    ./common/rdm/Pids.pb.h:17:2: error: #error This file was generated by an older version of protoc which is
     #error This file was generated by an older version of protoc which is
      ^
    ./common/rdm/Pids.pb.h:18:2: error: #error incompatible with your Protocol Buffer headers. Please
     #error incompatible with your Protocol Buffer headers.  Please
      ^
    ./common/rdm/Pids.pb.h:19:2: error: #error regenerate this file with a newer version of protoc.
     #error regenerate this file with a newer version of protoc.

> Then you should run `make clean to clean out the old protobuf files.

Reviewer 2: ```suggestion
Then you should run `make clean` to clean out the old protobuf files.
```

Optional
========

Doxygen Documentation
---------------------

There is also an option to build the doxygen documentation! To do so, you will need to install
[doxygen](http://www.doxygen.nl/manual/install.html) (available from package managers on all popular systems).

Once you have installed Doxygen you may need to run `./configure` in your source directory to generate the correct
makefile. To build the docs just use:

     make doxygen-doc

> You’ll have to run a webserver to get the experience. A simple way to do this is to navigate to `open-lighting/html` and

Reviewer 2: That will be your source directory's html folder...
